proposals/infra/INF-0004-inproc-outofproc-compiler-library-support.md

```diff
@@ -0,0 +1,113 @@
+considered protected.  An out of process design would move that concern to an
+isolated process.
+
+* In-process support using multiple threads have known thread-safety concerns
+for the clang compiler.  The clang compiler holds some state at compile time
+and some of that state could leak to other compilation sessions if it is not
+protected. This issue was also called out when an experiment called
+[llvm-buildozer](https://reviews.llvm.org/D86351) was created.  The author was
+interested in multiple threads re-entering the main compiler entry point in an
+effort to reduce compilation times. The effort showed that the unsafe bits
+could be fixed up and made thread safe allowing the experiment to be built.
+I have not checked to see the thread safe fixed bits are in the current llvm
+source.
+
+## Proposed solution
+
+### What could an out-of-process design look like?
+
+The architecture for an out of process design will behave in a similar way to
+the MSBuild design. The system functions as a Process Pool.  This allows the
+the compilation work to take advantage of systems that have multiple
+processors, or multiple-core processors. A separate compiler process
+is created for each available processor. For example, if the system has four
+processors, then four compiler processes are created.
+
+The process pool will be associated to an instance of the compiler library
+and will live as long as that instance is alive.  Compilation requests will
+be queued and the pool of processes that work through compilation requests.
+
+Communication between with the process pool will be done using a named pipe
+IPC mechanism. Pipe names will be unique to the process that is being
+communicated with. Results are communicated back over the IPC mechanism.
+
+### Error handling
+
+If the HLSL compiler encounters an error during compilation or the compiler
+process crashes, the rest of the compiler processes will continue on.
+Error information is communicated back over the IPC mechanism to the caller
+and the application will choose how to handle it.
+
+### What if in-process becomes a requirement at a later time?
+
+If an in process solution becomes a requirement, it can easily be built
+because the support code that performs compilation will have already been
+refactored out to be called by the out of process design.
+
+### Open Questions
+
+* Should the portions of the out of process architecture be built as a general
+purpose system for others in the llvm project to be able to use in their own
+systems?
+
+## Detailed design
+
+A detailed design of the out-of-process system has not been created.
+There is an interesting example of an approach for a general purpose system
+called 'Compiler Services'. This was presented as an option to support editors
+and work with compiler databases. The proposal was called
+[Clang C++ Services](https://github.com/chandlerc/llvm-designs/blob/master/ClangService.rst)
+
+## Resources
```

Comment:
The template has an "Alternatives Considered" option, which might be a good place to discuss why inproc-only or inproc+outproc were not chosen as the solution.

Reply:
ok.. Will do.

Reply:
Fixed